src/core/net/ip6.cpp

```diff
@@ -1492,6 +1502,66 @@ bool Ip6::IsOnLink(const Address &aAddress) const
     return rval;
 }
 
+void Ip6::UpdateBorderRoutingCounters(const Message &  aMessage,
+                                      const Header &   aHeader,
+                                      const Header *   aInnerHeader,
+                                      bool             aForwardHost,
+                                      bool             aForwardThread,
+                                      const OptionMpl &aMplOption)
+{
+    otPacketsAndBytes *counter = nullptr;
+
+    if (!Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetSource()) &&
+        !Get<Netif>().GetUnicastAddresses().ContainsMatching(aHeader.GetDestination()) &&
+        !aHeader.GetSource().IsLinkLocal() && !aHeader.GetDestination().IsLinkLocal() &&
+        aHeader.GetSource().GetPrefix() != Get<Mle::Mle>().GetMeshLocalPrefix() &&
+        aHeader.GetDestination().GetPrefix() != Get<Mle::Mle>().GetMeshLocalPrefix())
+    {
+        if (aForwardThread)
+        {
+            if (aHeader.GetDestination().IsMulticast())
+            {
+                // Do nothing. Inbound Multicast is handled later.
+            }
+            else
+            {
+                counter = &mBorderRoutingCounters.mInboundUnicast;
```

Comment:
@superwhd , I think the goal is to count the number of IP packets forwarded between infra link and thread (basically count packets being "border routed" between two networks).

I think the code here and the model of using `Ip6::HandleDatagram` and `aForwardThread` may not update the counter correctly and may over-count. 
- We can receive an IPv6 message over Thread radio which is passed to `HandleDatagram` and then we decide to forward it back into Thread (`aForwardThead` will be `true`).
- But then here we treat it as if it was from infra link and include in the counter.
- The most common example of this is when the device is the parent of some sleepy children, then other thread nodes will send the message destined for the sleepy child's address to the parent and then parent will forward it to sleepy child
    -  Parent of sleepy will respond to ddress query for all SED address and gives its own RLOC16. So other node will send their msg to the parent.